### PR TITLE
fix(auth): Fixes authentication conditions and logic with registries

### LINKED ISSUE #4157: Yarn writes registry.yarnpkg.com to lockfile when private scoped packages is from registry.npmjs.org
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
I have something like the following in my ~/.npmrc file:
```
@mycompany:registry=https://registry.npmjs.org/
//registry.npmjs.org/:_authToken=XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
```
This is necessary so I can login to my private repo and get the packages I need to build locally. 

When I add a new package like @mycompany/foo I would expect an entry in the `yarn.lock` file to look something like this:
```
"@mycompany/foo@^1.1.0":
  version "1.1.0"
  resolved "https://registry.npmjs.org/@mycompany/foo/-/foo-1.1.0.tgz#103034ba565c288eb781b1fd1065c9f05dfb90f9"
```
Yeah... not so much.  Instead it looks like this:
```
"@mycompany/foo@^1.1.0":
  version "1.1.0"
  resolved "https://registry.yarnpkg.com/@mycompany/foo/-/foo-1.1.0.tgz#103034ba565c288eb781b1fd1065c9f05dfb90f9"
```
Uncool yarn.  This causes my CI to break because it will try and dig up the packages at `registry.yarnpkg.com`.  Since this is a private registry the login will fail (since the registry is hosted on registry.npmjs.org) and thus it will cause other systems to fail to install packages.

Here is how to reproduce:

1. Setup a private repo on npm with some namespace
2. Set your .npmrc file to something like the one I provided above (except you replace the scope `mycompany` with your own scope, and set a valid auth token of course), and run `yarn add <your scoped package>`.  You can do this by doing `npm login --registry=https://registry.npmjs.org --scope=@<your scope name>`.  I would use yarn login but that doesn't work :(
3. Add a package from your scope using `yarn add`
4. Push a trivial update to that package in your scope.
5. Tell yarn to upgrade the package.

You will see that the URL gets changed in the lock file.

**Please mention your node.js, yarn and operating system version.**
node: 8.2.1
yarn: 0.28.4
OS: Fedora 26

### PR DESCRIPTION
**Pull request by KidkArolis**
### Summary

Supersedes #5162 (for now). Fixes #4157, #4451, #4672, #4119.

This PR brings 2 changes.

**Fix for private npm packages**

Fixes unauthed requests to yarnpk.com by treating yarnpkg.com as requests to npmjs.org in `isRequestToRegistry`.

**Feature - simpler support for multipath registries**

Before, if you wanted to use something like artifactory, Nexus or VSTS where meta url's differ from tarball urls, you had to configure like so:

```
@scope:registry=https://artifactory.example.com/api/npm/registry/
//artifactory.example.com/api/npm/registry/:_authToken=xxxx
custom-host-suffix=artifactory.example.com
```

But now you can do this instead:

```
@scope:registry=https://artifactory.example.com/api/npm/registry/
//artifactory.example.com/api/npm/registry/:_authToken=xxxx
//artifactory.example.com/api/tarballs/:_authToken=xxxx
```

This way we don't introduce a new option, we're just saying, aha, when accessing this kind of URL, make sure to authenticate it. I personally find it slightly more intuitive.

### Test plan

I've added a new test case to `requestIsToRegistry` unit test.

I've redone the request authorisation tests. It's a small hit to the ergonomics of the tests (harder to comment some of them or use skip/only feature). But compacting them all from code to data means we can be a lot more thorough with all the different cases. The intention was to preserve all of the existing test cases, but add many more. In particular, here are the different common scenarios I tested:

1. using npm as default registry and using private registry for scoped packages
2. using scoped packages in both npm and private registry
3. using authenticated npm and using private registry for scoped packages
4. using npm with always-auth and using private registry for scoped packages
5. using private registry as default registry and using scoped packages on npm registry

And then also tested

6. registry url and request url path sensitivity
7. using custom-host-suffix for registries where pathnames for meta and tarballs differ
8. using multiple authToken entries for registries where pathnames for meta and tarballs differ

The PR is split into somewhat independent commits if that helps.

### CODE REVIEW

**Comment by arcanis** [comment]
I'm going to go ahead and merge it for the release, but please check that the function is required :)

**Comment by JoostK** [comment]
@KidkArolis Is there a specific reason why only `_authToken` is considered, but not `_auth` (Basic) or `username`/`_password`? I'd like to be able to use the latter two authentication options as well.

**Comment by KidkArolis** [comment]
@JoostK Basic auth should work fine, it's handled in `getAuth` here: https://github.com/yarnpkg/yarn/blob/7fcf4dd4d95cdfa8856cea8a5daa29d48eb2d70b/src/registries/npm-registry.js#L338-L344. I guess it's just not been covered in these functional tests. I say that's fine until some reports an issue where it's not working.

**Comment by JoostK** [comment]
@KidkArolis I was referring to `requestNeedsAuth` as introduced in this PR, it only checks for `_authToken`.

File: src/registries/npm-registry.js [hunk 1]
```diff
@@ -100,21 +113,18 @@ export default class NpmRegistry extends Registry {
...
-    const registryParsed = url.parse(normalizedRegistryUrl);
-    const requestHost = requestParsed.host || '';
-    const registryHost = registryParsed.host || '';
-    const requestPath = requestParsed.path || '';
-    const registryPath = registryParsed.path || '';
+    const request = urlParts(requestUrl);
+    const registry = urlParts(registryUrl);
     const customHostSuffix = this.getRegistryOrGlobalOption(registryUrl, 'custom-host-suffix');
 
+    const requestToRegistryHost = () => request.host === registry.host;
```
[the remark this answers is not in the document]
Comment (KidkArolis) [reply src/registries/npm-registry.js]: Yeah, this is a bit silly. I might have been trying to "optimise" by avoiding expression evaluation until needed, but it's just string comparisons and perhaps creating functions inline is more expensive.

In any case - refactored this function slightly for overall readability:
https://github.com/yarnpkg/yarn/pull/5321

**Comment by KidkArolis** [comment]
Ah, right! Hm, I suppose I could add that for completeness.

You can still use username/password if you specify `custom-host-suffix` for your registry.

But I suppose this PR is proposing to avoid using `custom-host-suffix` in favor of configuring multiple paths of the same registry. I'll see if it's easy to add.

**Comment by KidkArolis** [comment]
@JoostK Well that was easy: https://github.com/yarnpkg/yarn/pull/5322.

**Comment by JoostK** [comment]
@KidkArolis Thanks! I wasn't aware of `custom-host-suffix`, totally glanced over it in your example in the description.

I prefer your approach with this PR as it is far more intuitive to configure and also more generic. In fact, I don't see a need for `custom-host-suffix` anymore.

Perhaps renaming the `isToRegistry` variable to `needsAuth` would also clear things up, it looks like a leftover now:
```js
const isToRegistry = this.isRequestToRegistry(requestUrl, registry) || this.requestNeedsAuth(requestUrl);
```

**Comment by KidkArolis** [comment]
There are a lot of conditions that are taken into account before we call `getAuth`:

```js
const isToRegistry = this.isRequestToRegistry(requestUrl, registry) || this.requestNeedsAuth(requestUrl);

    // this.token must be checked to account for publish requests on non-scopped packages
    if (this.token || (isToRegistry && (alwaysAuth || this.isScopedPackage(packageIdent)))) {
      const authorization = this.getAuth(packageIdent);
      if (authorization) {
        headers.authorization = authorization;
      }
    }
```

All of these conditions ultimately decide if the request `needsAuth`..

So the `isToRegistry` variable is still more meaningful in that it's more specific, it says "is this request to the registry or somewhere else".. But you're right in that `this.requestNeedsAuth()` doesn't quite fit in there..

There must be a way to rewrite this block into something better, but oh my, not sure how..

Maybe something like:

```js
const tokenInUse = !!this.token
const isToRegistry = this.isRequestToRegistry(requestUrl, registry)
const isScopedPackage = this.isScopedPackage(packageIdent)
const registryAuthRequired = this.requestNeedsAuth(requestUrl);

const requestNeedsAuth = (isToRegistry || registryAuthRequired) && (alwaysAuth || isScopedPackage)

if (tokenInUse || requestNeedsAuth) {
```

I'm just worried about making these expressions no longer lazy, could impact perf if calcs done for each package.

Maybe the only issue is the naming of `this.requestNeedsAuth`? It is actually a different way to check if the request is being made to some registry in the config. So maybe I need to pull it into `isRequestToRegistry` logic and `||` it there as `this.requestMatchesConfiguredRegistry` or smth.

**Pull request merged**